docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,284 @@
+  mark ourselves as the losing gateway etc.
+- Otherwise we process message normally and at the end update the twin as we currently do.
+
+#### Alternative
+
+- Use the Function to do both operations: idea was discarded because of a connection switch
+  - update the device twin frame counter to 0. As frame counter is a reported property it needs to be
+    changed via a DeviceClient that would cause a connection switch.
+  - Clear or update the cache entry with frame counter down and up to 0.
+  - Returns the result to the LNS: whether it was the winning or losing one
+  - LNS reacts as described in the [main data flow section](#main-data-message-flow)
+
+### Handling Class C downstream messages
+
+For class C devices we can send C2D messages using a Direct Method that could (one-off) steal the
+active connection. When the Direct method is invoked via the portal on an LNS, we should check if we
+are the connection holder for that device and if not drop the message.
+
+#### Should we delay on the LNS itself or on the Function?
+
+We considered using a delay on the Function rather than on the LNS itself. We decided against this approach
+because of the following disadvantages:
+
+- Observability: potentially we are messing up the measurements of the Function duration for the
+  LNSs that are not owning the connection. Could be documented/worked-around.
+- Keeps the HTTP connection between the LNS-Function open for more time.
+
+For the sake of completeness a scenario when it's better that the Function implements the delay is the following:
+
+- LNS1 is the preferred LNS. LNS2 is out of range.
+- LNS2 becomes in range and receives a message with a higher frame counter. It does not know that its
+  not the winning LNS and contacts immediately the Function. The Function awards it the winning LNS
+  and LNS1 loses the connection without having a chance to keep it.
+
+LNS2 would need to fetch the device twin so it is likely to lose the race to LNS1 but even if it
+does not we accept the possibility that there is potentially a one-off connection switch (but not a
+ping pong because LNS1 will stop retrying).
+
+A potential advantage of delaying on the LNS is that we can dynamically (based on how long we took in
+previous steps) choose the delay amount before contacting the Function, so that we have higher
+chances of not missing the window.
+  
+#### Delay amount configuration
+
+The delay amount should be configurable to allow users to customize behavior for their scenarios.
+During load testing we tested with 400ms but smaller values should be tried as well. If 0ms are
+specified the stickiness feature is disabled which means potential connection switching.
+
+## Other candidates considered
+
+### Single point of connection handling on LoRaDevice
+
+The changes presented from Atif: would also ensure by design that new code does not open more
```

Comment:
That's weird to put under "Other candidates considered" - I would put this under a related section to indicate that this is additional.

Reply:
Agree, moved to a related section